src/advanced_inv.cpp

```diff
@@ -1365,111 +1428,99 @@ void advanced_inventory::display()
-                        for( auto &elem : moving_items ) {
-                            assert( !elem.is_null() );
-                            if( chargeback ) {
-                                g->u.i_add( elem );
-                            } else if( add_item( destarea, elem ) ) {
-                                ++moved;
-                            } else {
-                                g->u.i_add( elem );
-                                chargeback = true;
-                            }
-                        }
-                        if( moved == 0 ) {
-                            continue;
+                    std::list<item> moving_items;
+                    if(srcarea == AIM_INVENTORY) {
+                        moving_items = g->u.inv.reduce_stack(idx, amount_to_move);
+                    } else if(srcarea == AIM_WORN) {
+                        std::vector<item> mv;
+                        g->u.takeoff(idx, false, &mv);
+                        std::copy(mv.begin(), mv.end(), std::back_inserter(moving_items));
+                    }
+                    int items_left = 0, moved = 0;
+                    for(auto &elem : moving_items) {
+                        assert(!elem.is_null());
+                        items_left = add_item(destarea, elem);
+                        if(items_left > 0) {
+                            // chargeback the items if adding them failed
+                            add_item(srcarea, elem, items_left);
+                        } else {
+                            ++moved;
                         }
                     }
+                    if(moved == 0) {
+                        continue;
+                    }
                 }
             } else {
                 // from map/vehicle: add the item to the destination, if that worked,
                 // remove it from the source, else continue.
-                // TODO: move partial stack with items that are not counted by charges
-                item new_item( *sitem->it );
-                if( by_charges ) {
-                    new_item.charges = amount_to_move;
-                }
-                if( !add_item( destarea, new_item ) ) {
-                    // make sure to redraw the info window, so 
-                    // the user can see why it wasn't equipped
-                    if(destarea == AIM_WORN) {
-                        redraw = 1;
+                // create a new copy of the old item being manipulated
+                item new_item(*sitem->items.front());
+                // add the item, and note any items that might be leftover
+                int items_left = add_item(destarea, new_item, (by_charges) ? 1 : amount_to_move);
+                // only remove item or charges if the add succeeded
+                if(items_left == 0) {
+                    if(by_charges) {
+                        // set the new item's charge amount
+                        new_item.charges = amount_to_move;
```

Comment:
This should probably be done *before* the item is added. Currently it has no effect as `new_item` is not used from this point on.